src/pyhf/modifiers/shapefactor.py

```diff
@@ -62,64 +63,71 @@ def __init__(self, shapefactor_mods, pdfconfig, mega_mods):
-            default_backend.tolist(
-                default_backend.gather(
-                    default_backend.astensor(
-                        indices + [0] * (max_nbins - len(indices)), dtype='int'
-                    ),
-                    global_concatenated_bin_indices,
-                )
-            )
-            for indices in self._shapefactor_indices
+            [[j for c in pdfconfig.channels for j in range(pdfconfig.channel_nbins[c])]]
         ]
 
+        self._access_field = default_backend.tile(
+            global_concatenated_bin_indices, (len(pnames), self.batch_size or 1, 1)
+        )
+        # access field is shape (sys, batch, globalbin)
+        for s, syst_access in enumerate(self._access_field):
+            for t, batch_access in enumerate(syst_access):
+                selection = self.parameters_helper.index_selection[s][t]
+                for b, bin_access in enumerate(batch_access):
+                    self._access_field[s, t, b] = (
+                        selection[bin_access] if bin_access < len(selection) else 0
+                    )
+
         self._precompute()
         events.subscribe('tensorlib_changed')(self._precompute)
 
     def _precompute(self):
-        if not self._shapefactor_indices:
+        if not self.parameters_helper.index_selection:
             return
         tensorlib, _ = get_backend()
         self.shapefactor_mask = tensorlib.astensor(self._shapefactor_mask)
+        self.shapefactor_mask = tensorlib.tile(
+            self.shapefactor_mask, (1, 1, self.batch_size or 1, 1)
+        )
+        self.access_field = tensorlib.astensor(self._access_field, dtype='int')
+
         self.shapefactor_default = tensorlib.ones(
             tensorlib.shape(self.shapefactor_mask)
         )
-        self.shapefactor_indices = tensorlib.astensor(
-            self._shapefactor_indices, dtype='int'
-        )
         self.sample_ones = tensorlib.ones(tensorlib.shape(self.shapefactor_mask)[1])
-        self.alpha_ones = tensorlib.ones([1])
 
     def apply(self, pars):
         '''
         Returns:
             modification tensor: Shape (n_modifiers, n_global_samples, n_alphas, n_global_bin)
         '''
-        if not self._shapefactor_indices:
+        if not self.parameters_helper.index_selection:
             return
+
         tensorlib, _ = get_backend()
-        shapefactors = tensorlib.gather(pars, self.shapefactor_indices)
+        pars = tensorlib.astensor(pars)
+        if self.batch_size is None:
+            batched_pars = tensorlib.reshape(pars, (1,) + tensorlib.shape(pars))
+        else:
+            batched_pars = pars
+
+        flat_pars = tensorlib.reshape(batched_pars, (-1,))
+        shapefactors = tensorlib.gather(flat_pars, self.access_field)
         results_shapefactor = tensorlib.einsum(
-            's,a,mb->msab', self.sample_ones, self.alpha_ones, shapefactors
+            'yab,s->ysab', shapefactors, self.sample_ones
```

Comment:
`y`?

Reply:
`yab`

* `y`: counter for instance of staterror
* `a`: batch direction
* `b`: global bin counter

same as below, if you prefer another letter lmk

Reply:
why not `m` for modifier?